pennylane/workflow/qnode.py

```diff
@@ -1080,22 +897,29 @@ def _execution_component(self, args: tuple, kwargs: dict, override_shots) -> qml
 
         execute_kwargs = copy.copy(self.execute_kwargs)
         mcm_config = copy.copy(execute_kwargs["mcm_config"])
-        finite_shots = _get_device_shots(self.device) if override_shots is False else override_shots
+        finite_shots = (lambda dev: dev.shots) if override_shots is False else override_shots
+
         if not finite_shots:
             mcm_config.postselect_mode = None
             if mcm_config.mcm_method in ("one-shot", "tree-traversal"):
                 raise ValueError(
                     f"Cannot use the '{mcm_config.mcm_method}' method for mid-circuit measurements with analytic mode."
                 )
+
         if mcm_config.mcm_method == "single-branch-statistics":
             raise ValueError("Cannot use mcm_method='single-branch-statistics' without qml.qjit.")
 
         full_transform_program = qml.transforms.core.TransformProgram(self.transform_program)
         inner_transform_program = qml.transforms.core.TransformProgram()
         config = None
 
-        if isinstance(self.device, qml.devices.Device):
+        config = _make_execution_config(self, self.gradient_fn)
+        device_transform_program, config = self.device.preprocess(execution_config=config)
 
+        if config.use_device_gradient:
+            full_transform_program += device_transform_program
+        else:
+            inner_transform_program += device_transform_program

```

Comment:
I might be missing something, but it looks like lines 923-929 need to be deleted.